File: code/java/ch18/src/main/java/com/github/zxh0/luago/compiler/parser/ExpParser.java
package com.github.zxh0.luago.compiler.parser;

import com.github.zxh0.luago.compiler.ast.Block;
import com.github.zxh0.luago.compiler.ast.Exp;
import com.github.zxh0.luago.compiler.ast.exps.*;
import com.github.zxh0.luago.compiler.lexer.Lexer;
import com.github.zxh0.luago.compiler.lexer.Token;
import com.github.zxh0.luago.compiler.lexer.TokenKind;
import com.github.zxh0.luago.number.LuaNumber;

import java.util.ArrayList;
import java.util.List;

import static com.github.zxh0.luago.compiler.lexer.TokenKind.*;
import static com.github.zxh0.luago.compiler.parser.BlockParser.parseBlock;
import static com.github.zxh0.luago.compiler.parser.Optimizer.*;
import static com.github.zxh0.luago.compiler.parser.PrefixExpParser.parsePrefixExp;

class ExpParser {

    // explist ::= exp {‘,’ exp}
    static List<Exp> parseExpList(Lexer lexer) {
        List <Exp> exps = new ArrayList<>();
        exps.add(parseExp(lexer));
        while (lexer.LookAhead() == TOKEN_SEP_COMMA) {
            lexer.nextToken();
            exps.add(parseExp(lexer));
        }
        return exps;
    }

    /*
    exp ::=  nil | false | true | Numeral | LiteralString | ‘...’ | functiondef |
         prefixexp | tableconstructor | exp binop exp | unop exp
    */
    /*
    exp   ::= exp12
    exp12 ::= exp11 {or exp11}
    exp11 ::= exp10 {and exp10}
    exp10 ::= exp9 {(‘<’ | ‘>’ | ‘<=’ | ‘>=’ | ‘~=’ | ‘==’) exp9}
    exp9  ::= exp8 {‘|’ exp8}
    exp8  ::= exp7 {‘~’ exp7}
    exp7  ::= exp6 {‘&’ exp6}
    exp6  ::= exp5 {(‘<<’ | ‘>>’) exp5}
    exp5  ::= exp4 {‘..’ exp4}
    exp4  ::= exp3 {(‘+’ | ‘-’ | ‘*’ | ‘/’ | ‘//’ | ‘%’) exp3}
    exp2  ::= {(‘not’ | ‘#’ | ‘-’ | ‘~’)} exp1
    exp1  ::= exp0 {‘^’ exp2}
    exp0  ::= nil | false | true | Numeral | LiteralString
            | ‘...’ | functiondef | prefixexp | tableconstructor
    */
    static Exp parseExp(Lexer lexer) {
        return parseExp12(lexer);
    }


    // x or y
    private static Exp parseExp12(Lexer lexer) {
        Exp exp = parseExp11(lexer);
        while (lexer.LookAhead() == TOKEN_OP_OR) {
            Token op = lexer.nextToken();
            BinopExp lor = new BinopExp(op, exp, parseExp11(lexer));
            exp = optimizeLogicalOr(lor);
        }
        return exp;
    }

    // x and y
    private static Exp parseExp11(Lexer lexer) {
        Exp exp = parseExp10(lexer);
        while (lexer.LookAhead() == TOKEN_OP_AND) {
            Token op = lexer.nextToken();
            BinopExp land = new BinopExp(op, exp, parseExp10(lexer));
            exp = optimizeLogicalAnd(land);
        }
        return exp;
    }

    // compare
    private static Exp parseExp10(Lexer lexer) {
        Exp exp = parseExp9(lexer);
        while (true) {
            switch (lexer.LookAhead()) {
                case TOKEN_OP_LT:
                case TOKEN_OP_GT:
                case TOKEN_OP_NE:
                case TOKEN_OP_LE:
                case TOKEN_OP_GE:
                case TOKEN_OP_EQ:
                    Token op = lexer.nextToken();
                    exp = new BinopExp(op, exp, parseExp9(lexer));
                    break;
                default:
                    return exp;
            }
        }
    }

    // x | y
    private static Exp parseExp9(Lexer lexer) {
        Exp exp = parseExp8(lexer);
        while (lexer.LookAhead() == TOKEN_OP_BOR) {
            Token op = lexer.nextToken();
            BinopExp bor = new BinopExp(op, exp, parseExp8(lexer));
            exp = optimizeBitwiseBinaryOp(bor);
        }
        return exp;
    }

    // x ~ y
    private static Exp parseExp8(Lexer lexer) {
        Exp exp = parseExp7(lexer);
        while (lexer.LookAhead() == TOKEN_OP_WAVE) {
            Token op = lexer.nextToken();
            BinopExp bxor = new BinopExp(op, exp, parseExp7(lexer));
            exp = optimizeBitwiseBinaryOp(bxor);
        }
        return exp;
    }

    // x & y
    private static Exp parseExp7(Lexer lexer) {
        Exp exp = parseExp6(lexer);
        while (lexer.LookAhead() == TOKEN_OP_BAND) {
            Token op = lexer.nextToken();
            BinopExp band = new BinopExp(op, exp, parseExp6(lexer));
            exp = optimizeBitwiseBinaryOp(band);
        }
        return exp;
    }

    // shift
    private static Exp parseExp6(Lexer lexer) {
        Exp exp = parseExp5(lexer);
        while (true) {
            switch (lexer.LookAhead()) {
                case TOKEN_OP_SHL:
                case TOKEN_OP_SHR:
                    Token op = lexer.nextToken();
                    BinopExp shx = new BinopExp(op, exp, parseExp5(lexer));
                    exp = optimizeBitwiseBinaryOp(shx);
                    break;
                default:
                    return exp;
            }
        }
    }

    // a .. b
    private static Exp parseExp5(Lexer lexer) {
        Exp exp = parseExp4(lexer);
        if (lexer.LookAhead() != TOKEN_OP_CONCAT) {
            return exp;
        }

        List<Exp> exps = new ArrayList<>();
        exps.add(exp);
        int line = 0;
        while (lexer.LookAhead() == TOKEN_OP_CONCAT) {
            line = lexer.nextToken().getLine();
            exps.add(parseExp4(lexer));
        }
        return new ConcatExp(line, exps);
    }

    // x +/- y
    private static Exp parseExp4(Lexer lexer) {
        Exp exp = parseExp3(lexer);
        while (true) {
            switch (lexer.LookAhead()) {
                case TOKEN_OP_ADD:
                case TOKEN_OP_MINUS:
                    Token op = lexer.nextToken();
                    BinopExp arith = new BinopExp(op, exp, parseExp3(lexer));
                    exp = optimizeArithBinaryOp(arith);
                    break;
                default:
                    return exp;
            }
        }
    }

    // *, %, /, //
    private static Exp parseExp3(Lexer lexer) {
        Exp exp = parseExp2(lexer);
        while (true) {
            switch (lexer.LookAhead()) {
                case TOKEN_OP_MUL:
                case TOKEN_OP_MOD:
                case TOKEN_OP_DIV:
                case TOKEN_OP_IDIV:
                    Token op = lexer.nextToken();
                    BinopExp arith = new BinopExp(op, exp, parseExp2(lexer));
                    exp = optimizeArithBinaryOp(arith);
                    break;
                default:
                    return exp;
            }
        }
    }

    // unary
    private static Exp parseExp2(Lexer lexer) {
        switch (lexer.LookAhead()) {
            case TOKEN_OP_MINUS:
            case TOKEN_OP_WAVE:
            case TOKEN_OP_LEN:
            case TOKEN_OP_NOT:
                Token op = lexer.nextToken();
                UnopExp exp = new UnopExp(op, parseExp2(lexer));
            return optimizeUnaryOp(exp);
        }
        return parseExp1(lexer);
    }

    // x ^ y
    private static Exp parseExp1(Lexer lexer) { // pow is right associative
        Exp exp = parseExp0(lexer);
        if (lexer.LookAhead() == TOKEN_OP_POW) {
            Token op = lexer.nextToken();
            exp = new BinopExp(op, exp, parseExp2(lexer));
        }
        return optimizePow(exp);
    }

    private static Exp parseExp0(Lexer lexer) {
        switch (lexer.LookAhead()) {
            case TOKEN_VARARG: // ...
                return new VarargExp(lexer.nextToken().getLine());
            case TOKEN_KW_NIL: // nil
                return new NilExp(lexer.nextToken().getLine());
            case TOKEN_KW_TRUE: // true
                return new TrueExp(lexer.nextToken().getLine());
            case TOKEN_KW_FALSE: // false
                return new FalseExp(lexer.nextToken().getLine());
            case TOKEN_STRING: // LiteralString
                return new StringExp(lexer.nextToken());
            case TOKEN_NUMBER: // Numeral
                return parseNumberExp(lexer);
            case TOKEN_SEP_LCURLY: // tableconstructor
                return parseTableConstructorExp(lexer);
            case TOKEN_KW_FUNCTION: // functiondef
                lexer.nextToken();
                return parseFuncDefExp(lexer);
            default: // prefixexp
                return parsePrefixExp(lexer);
        }
    }

    private static Exp parseNumberExp(Lexer lexer) {
        Token token = lexer.nextToken();
        Long i = LuaNumber.parseInteger(token.getValue());
        if (i != null) {
            return new IntegerExp(token.getLine(), i);
        }
        Double f = LuaNumber.parseFloat(token.getValue());
        if (f != null) {
            return new FloatExp(token.getLine(), f);
        }
        throw new RuntimeException("not a number: " + token);
    }

    // functiondef ::= function funcbody
    // funcbody ::= ‘(’ [parlist] ‘)’ block end
    static FuncDefExp parseFuncDefExp(Lexer lexer) {
        int line = lexer.line();                    // function
        lexer.nextTokenOfKind(TOKEN_SEP_LPAREN);    // (
        List<String> parList = parseParList(lexer); // [parlist]
        lexer.nextTokenOfKind(TOKEN_SEP_RPAREN);    // )
        Block block = parseBlock(lexer);            // block
        lexer.nextTokenOfKind(TOKEN_KW_END);        // end
        int lastLine = lexer.line();

        FuncDefExp fdExp = new FuncDefExp();
        fdExp.setLine(line);
        fdExp.setLastLine(lastLine);
        fdExp.setIsVararg(parList.remove("..."));
        fdExp.setParList(parList);
        fdExp.setBlock(block);
        return fdExp;
    }

    // [parlist]
    // parlist ::= namelist [‘,’ ‘...’] | ‘...’
    private static List<String> parseParList(Lexer lexer) {
        List<String> names = new ArrayList<>();

        switch (lexer.LookAhead()) {
            case TOKEN_SEP_RPAREN:
                return names;
            case TOKEN_VARARG:
                lexer.nextToken();
                names.add("...");
                return names;
        }

        names.add(lexer.nextIdentifier().getValue());
        while (lexer.LookAhead() == TOKEN_SEP_COMMA) {
            lexer.nextToken();
            if (lexer.LookAhead() == TOKEN_IDENTIFIER) {
                names.add(lexer.nextIdentifier().getValue());
            } else {
                lexer.nextTokenOfKind(TOKEN_VARARG);
                names.add("...");
                break;
            }
        }

        return names;
    }

    // tableconstructor ::= ‘{’ [fieldlist] ‘}’
    static TableConstructorExp parseTableConstructorExp(Lexer lexer) {
        TableConstructorExp tcExp = new TableConstructorExp();
        tcExp.setLine(lexer.line());
        lexer.nextTokenOfKind(TOKEN_SEP_LCURLY); // {
        parseFieldList(lexer, tcExp);            // [fieldlist]
        lexer.nextTokenOfKind(TOKEN_SEP_RCURLY); // }
        tcExp.setLastLine(lexer.line());
        return tcExp;
    }

    // fieldlist ::= field {fieldsep field} [fieldsep]
    private static void parseFieldList(Lexer lexer, TableConstructorExp tcExp) {
        if (lexer.LookAhead() != TOKEN_SEP_RCURLY) {
            parseField(lexer, tcExp);

            while (isFieldSep(lexer.LookAhead())) {
                lexer.nextToken();
                if (lexer.LookAhead() != TOKEN_SEP_RCURLY) {
                    parseField(lexer, tcExp);
                } else {
                    break;
                }
            }
        }
    }

    // fieldsep ::= ‘,’ | ‘;’
    private static boolean isFieldSep(TokenKind kind) {
        return kind == TOKEN_SEP_COMMA || kind == TOKEN_SEP_SEMI;
    }

    // field ::= ‘[’ exp ‘]’ ‘=’ exp | Name ‘=’ exp | exp
    private static void parseField(Lexer lexer, TableConstructorExp tcExp) {
        if (lexer.LookAhead() == TOKEN_SEP_LBRACK) {
            lexer.nextToken();                       // [
            tcExp.addKey(parseExp(lexer));           // exp
            lexer.nextTokenOfKind(TOKEN_SEP_RBRACK); // ]
            lexer.nextTokenOfKind(TOKEN_OP_ASSIGN);  // =
            tcExp.addVal(parseExp(lexer));           // exp
            return;
        }

        Exp exp = parseExp(lexer);
        if (exp instanceof NameExp) {
            if (lexer.LookAhead() == TOKEN_OP_ASSIGN) {
                // Name ‘=’ exp => ‘[’ LiteralString ‘]’ = exp
                tcExp.addKey(new StringExp(exp.getLine(), ((NameExp) exp).getName()));
                lexer.nextToken();
                tcExp.addVal(parseExp(lexer));
                return;
            }
        }

        tcExp.addKey(null);
        tcExp.addVal(exp);
    }

}

File: code/python/ch15_16/src/exp_parser.py
from lua_token import TokenKind
import lua_exp
from optimizer import Optimizer
from lua_value import LuaValue


class ExpParser:
    @staticmethod
    def parse_exp_list(lexer):
        exps = [ExpParser.parse_exp(lexer)]
        while lexer.look_ahead() == TokenKind.SEP_COMMA:
            lexer.get_next_token()
            exps.append(ExpParser.parse_exp(lexer))
        return exps

    @staticmethod
    def parse_exp(lexer):
        return ExpParser.parse_exp12(lexer)

    # x or y
    @staticmethod
    def parse_exp12(lexer):
        exp = ExpParser.parse_exp11(lexer)
        while lexer.look_ahead() == TokenKind.OP_OR:
            line, op, _ = lexer.get_next_token()
            lor = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp11(lexer))
            exp = Optimizer.optimize_logical_or(lor)
        return exp

    # x and y
    @staticmethod
    def parse_exp11(lexer):
        exp = ExpParser.parse_exp10(lexer)
        while lexer.look_ahead() == TokenKind.OP_AND:
            line, op, _ = lexer.get_next_token()
            lor = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp10(lexer))
            exp = Optimizer.optimize_logical_and(lor)
        return exp

    # compare
    @staticmethod
    def parse_exp10(lexer):
        exp = ExpParser.parse_exp9(lexer)
        while True:
            kind = lexer.look_ahead()
            if kind in (TokenKind.OP_LT, TokenKind.OP_GT, TokenKind.OP_NE,
                        TokenKind.OP_LE, TokenKind.OP_GE, TokenKind.OP_EQ):
                line, op, _ = lexer.get_next_token()
                exp = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp9(lexer))
            else:
                return exp

    # x | y
    @staticmethod
    def parse_exp9(lexer):
        exp = ExpParser.parse_exp8(lexer)
        while lexer.look_ahead() == TokenKind.OP_BOR:
            line, op, _ = lexer.get_next_token()
            bor = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp8(lexer))
            exp = Optimizer.optimize_bitwise_binary_op(bor)
        return exp

    # x ~ y
    @staticmethod
    def parse_exp8(lexer):
        exp = ExpParser.parse_exp7(lexer)
        while lexer.look_ahead() == TokenKind.OP_BXOR:
            line, op, _ = lexer.get_next_token()
            bor = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp8(lexer))
            exp = Optimizer.optimize_bitwise_binary_op(bor)
        return exp

    # x & y
    @staticmethod
    def parse_exp7(lexer):
        exp = ExpParser.parse_exp6(lexer)
        while lexer.look_ahead() == TokenKind.OP_BAND:
            line, op, _ = lexer.get_next_token()
            bor = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp8(lexer))
            exp = Optimizer.optimize_bitwise_binary_op(bor)
        return exp

    # shift
    @staticmethod
    def parse_exp6(lexer):
        exp = ExpParser.parse_exp5(lexer)
        if lexer.look_ahead() in (TokenKind.OP_SHL, TokenKind.OP_SHR):
            line, op, _ = lexer.get_next_token()
            shx = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp5(lexer))
            exp = Optimizer.optimize_bitwise_binary_op(shx)
        else:
            return exp
        return exp

    # a .. b
    @staticmethod
    def parse_exp5(lexer):
        exp = ExpParser.parse_exp4(lexer)
        if lexer.look_ahead() != TokenKind.OP_CONCAT:
            return exp

        line = 0
        exps = []
        while lexer.look_ahead() == TokenKind.OP_CONCAT:
            line, _, _ = lexer.get_next_token()
            exps.append(ExpParser.parse_exp4(lexer))
        return lua_exp.ConcatExp(line, exps)

    # x +/- y
    @staticmethod
    def parse_exp4(lexer):
        exp = ExpParser.parse_exp3(lexer)
        while True:
            if lexer.look_ahead() in (TokenKind.OP_ADD, TokenKind.OP_SUB):
                line, op, _ = lexer.get_next_token()
                arith = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp3(lexer))
                exp = Optimizer.optimize_arith_binary_op(arith)
            else:
                break
        return exp

    # *, %, /, //
    @staticmethod
    def parse_exp3(lexer):
        exp = ExpParser.parse_exp2(lexer)
        while True:
            if lexer.look_ahead() in (TokenKind.OP_MUL, TokenKind.OP_MOD, TokenKind.OP_DIV, TokenKind.OP_IDIV):
                line, op, _ = lexer.get_next_token()
                arith = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp2(lexer))
                exp = Optimizer.optimize_arith_binary_op(arith)
            else:
                break
        return exp

    # unary
    @staticmethod
    def parse_exp2(lexer):
        if lexer.look_ahead() in (TokenKind.OP_UNM, TokenKind.OP_BNOT, TokenKind.OP_LEN, TokenKind.OP_NOT):
            line, op, _ = lexer.get_next_token()
            exp = lua_exp.UnopExp(line, op, ExpParser.parse_exp2(lexer))
            return Optimizer.optimize_unary_op(exp)
        return ExpParser.parse_exp1(lexer)

    # x ^ y
    @staticmethod
    def parse_exp1(lexer):
        exp = ExpParser.parse_exp0(lexer)
        if lexer.look_ahead() == TokenKind.OP_POW:
            line, op, _ = lexer.get_next_token()
            exp = lua_exp.BinopExp(line, op, exp, ExpParser.parse_exp2(lexer))
        return Optimizer.optimize_pow(exp)

    @staticmethod
    def parse_exp0(lexer):
        kind = lexer.look_ahead()
        if kind == TokenKind.VARARG:
            line, _, _ = lexer.get_next_token()
            return lua_exp.VarArgExp(line)
        if kind == TokenKind.KW_NIL:
            line, _, _ = lexer.get_next_token()
            return lua_exp.NilExp(line)
        if kind == TokenKind.KW_TRUE:
            line, _, _ = lexer.get_next_token()
            return lua_exp.TrueExp(line)
        if kind == TokenKind.KW_FALSE:
            line, _, _ = lexer.get_next_token()
            return lua_exp.FalseExp(line)
        if kind == TokenKind.STRING:
            line, _, token = lexer.get_next_token()
            return lua_exp.StringExp(line, token)
        if kind == TokenKind.NUMBER:
            return ExpParser.parse_number_exp(lexer)
        if kind == TokenKind.SEP_LCURLY:
            return ExpParser.parse_table_constructor_exp(lexer)
        if kind == TokenKind.KW_FUNCTION:
            lexer.get_next_token()
            return ExpParser.parse_func_def_exp(lexer)
        return ExpParser.parse_prefix_exp(lexer)

    @staticmethod
    def parse_number_exp(lexer):
        line, _, token = lexer.get_next_token()
        i = LuaValue.parse_integer(token)
        if i is not None:
            return lua_exp.IntegerExp(line, i)
        f = LuaValue.parse_float(token)
        if f is not None:
            return lua_exp.FloatExp(line, f)
        raise Exception('not a number: ' + token)

    # functiondef::= function funcbody
    # funcbody::= '(' [parlist] ')' block end
    @staticmethod
    def parse_func_def_exp(lexer):
        from parser import Parser
        line = lexer.get_line()
        lexer.get_next_token_of_kind(TokenKind.SEP_LPAREN)
        par_list, is_var_arg = ExpParser.parse_par_list(lexer)
        lexer.get_next_token_of_kind(TokenKind.SEP_RPAREN)
        block = Parser.parse_block(lexer)
        last_line, _ = lexer.get_next_token_of_kind(TokenKind.KW_END)
        return lua_exp.FuncDefExp(line, last_line, par_list, is_var_arg, block)

    # [parlist]
    # parlist ::= namelist [',' '...'] | '...'
    @staticmethod
    def parse_par_list(lexer):
        kind = lexer.look_ahead()
        if kind == TokenKind.SEP_RPAREN:
            return None, False
        if kind == TokenKind.VARARG:
            return None, True

        _, name = lexer.get_next_identifier()
        names = [name]
        is_var_arg = False
        while lexer.look_ahead() == TokenKind.SEP_COMMA:
            lexer.get_next_token()
            if lexer.look_ahead() == TokenKind.IDENTIFIER:
                _, name = lexer.get_next_identifier()
                names.append(name)
            else:
                lexer.get_next_token_of_kind(TokenKind.VARARG)
                is_var_arg = True
                break
        return names, is_var_arg

    # tableconstructor ::= '{' [fieldlist] '}'
    @staticmethod
    def parse_table_constructor_exp(lexer):
        line = lexer.get_line()
        lexer.get_next_token_of_kind(TokenKind.SEP_LCURLY)
        key_exps, val_exps = ExpParser.parse_field_list(lexer)
        lexer.get_next_token_of_kind(TokenKind.SEP_RCURLY)
        last_line = lexer.get_line()
        return lua_exp.TableConstructorExp(line, last_line, key_exps, val_exps)

    # fieldlist ::= field {fieldsep field} [fieldsep]
    @staticmethod
    def parse_field_list(lexer):
        ks = []
        vs = []
        if lexer.look_ahead != TokenKind.SEP_RCURLY:
            k, v = ExpParser.parse_field(lexer)
            ks.append(k)
            vs.append(v)

            while ExpParser.is_field_sep(lexer.look_ahead()):
                lexer.get_next_token()
                if lexer.look_ahead() != TokenKind.SEP_RCURLY:
                    k, v = ExpParser.parse_field(lexer)
                    ks.append(k)
                    vs.append(v)
                else:
                    break
        return ks, vs

    # fieldsep ::= ',' | ';'
    @staticmethod
    def is_field_sep(kind):
        return kind in (TokenKind.SEP_COMMA, TokenKind.SEP_SEMI)

    # field ::= '[' exp ']' '=' exp | Name '=' exp | exp
    @staticmethod
    def parse_field(lexer):
        if lexer.look_ahead() == TokenKind.SEP_LBRACK:
            lexer.get_next_token()
            k = ExpParser.parse_exp(lexer)
            lexer.get_next_token_of_kind(TokenKind.SEP_RBRACK)
            lexer.get_next_token_of_kind(TokenKind.OP_ASSIGN)
            v = ExpParser.parse_exp(lexer)
            return k, v

        exp = ExpParser.parse_exp(lexer)
        if isinstance(exp, lua_exp.NameExp):
            if lexer.look_ahead() == TokenKind.OP_ASSIGN:
                lexer.get_next_token()
                k = lua_exp.StringExp(exp.line, exp.name)
                v = ExpParser.parse_exp(lexer)
                return k, v

        return None, exp

    # prefixexp::= var | functioncall | '(' exp ')"
    # var::= Name | prefixexp '[' exp ']' | prefixexp '.' Name
    # functioncall ::=  prefixexp
    # args | prefixexp ':' Name
    # args

    # prefixexp::= Name
    #   | '(' exp ')'
    #   | prefixexp '[' exp ']'
    #   | prefixexp '.' Name
    #   | prefixexp[':' Name] args
    @staticmethod
    def parse_prefix_exp(lexer):
        if lexer.look_ahead() == TokenKind.IDENTIFIER:
            line, name = lexer.get_next_identifier()
            exp = lua_exp.NameExp(line, name)
        else:
            exp = ExpParser.parse_parens_exp(lexer)
        return ExpParser.finish_prefix_exp(lexer, exp)

    @staticmethod
    def parse_parens_exp(lexer):
        lexer.get_next_token_of_kind(TokenKind.SEP_LPAREN)
        exp = ExpParser.parse_exp(lexer)
        lexer.get_next_token_of_kind(TokenKind.SEP_RPAREN)

        if isinstance(exp, lua_exp.VarArgExp) or isinstance(exp, lua_exp.FuncCallExp) or \
                isinstance(exp, lua_exp.NameExp) or isinstance(exp, lua_exp.TableAccessExp):
            return lua_exp.ParensExp(exp)
        return exp

    @staticmethod
    def finish_prefix_exp(lexer, exp):
        while True:
            kind = lexer.look_ahead()
            if kind == TokenKind.SEP_LBRACK:
                lexer.get_next_token()
                key_exp = ExpParser.parse_exp(lexer)
                lexer.get_next_token_of_kind(TokenKind.SEP_RBRACK)
                exp = lua_exp.TableAccessExp(lexer.get_line(), exp, key_exp)
            elif kind == TokenKind.SEP_DOT:
                lexer.get_next_token()
                line, name = lexer.get_next_identifier()
                key_exp = lua_exp.StringExp(line, name)
                exp = lua_exp.TableAccessExp(line, exp, key_exp)
            elif kind in (TokenKind.SEP_COLON, TokenKind.SEP_LPAREN, TokenKind.SEP_LCURLY, TokenKind.STRING):
                exp = ExpParser.finish_func_call_exp(lexer, exp)
            return exp

    # functioncall ::= prefixexp args | prefixexp ':' Name args
    @staticmethod
    def finish_func_call_exp(lexer, prefix_exp):
        name_exp = ExpParser.parse_name_exp(lexer)
        line = lexer.get_line()
        args = ExpParser.parse_args(lexer)
        last_line = lexer.get_line()
        return lua_exp.FuncCallExp(line, last_line, prefix_exp, name_exp, args)

    @staticmethod
    def parse_name_exp(lexer):
        if lexer.look_ahead() == TokenKind.SEP_COLON:
            lexer.get_next_token()
            line, name = lexer.get_next_identifier()
            return lua_exp.StringExp(line, name)
        return None

    # args ::= '(' [explist] ')' | tableconstructor | LitreralString
    @staticmethod
    def parse_args(lexer):
        args = []
        kind = lexer.look_ahead()
        if kind == TokenKind.SEP_LPAREN:    # '(' [explist] ')'
            lexer.get_next_token()
            if lexer.look_ahead() != TokenKind.SEP_RPAREN:
                args = ExpParser.parse_exp_list(lexer)
            lexer.get_next_token_of_kind(TokenKind.SEP_RPAREN)
        elif kind == TokenKind.SEP_LCURLY:  # '{' [fieldlist] '}'
            args = [ExpParser.parse_table_constructor_exp(lexer)]
        else:                               # LiteralString
            line, s = lexer.get_next_token_of_kind(TokenKind.STRING)
            args = [lua_exp.StringExp(line, s)]
        return args
